File: turtle_mover/scripts/turtle_mover.py
import math
import rospy
from geometry_msgs.msg import Twist
from turtlesim.msg import Pose 

# publisher
pub = rospy.Publisher('turtle1/cmd_vel', Twist, queue_size=10)

# robot starting point
initial_x = -1 
initial_y = -1
print_pos = False # print result


def move(distance:float, speed: float):
    vel_msg = Twist()
    vel_msg.linear.x = speed
    current_x = 0
    start_time = rospy.Time.now().to_sec()
    rate = rospy.Rate(30)

    while current_x < distance:
        pub.publish(vel_msg) #publish message
        current_time = rospy.Time.now().to_sec()
        current_x = (current_time - start_time) * speed
        rospy.loginfo("Current x position: %f", current_x)
        rate.sleep()

    # stop the robot
    vel_msg.linear.x = 0
    pub.publish(vel_msg)
    
def rotate(angle: float, speed: float):
    vel_msg = Twist()
    angular_velocity = speed * 2 * math.pi / 360
    vel_msg.angular.z = angular_velocity
    current_angular_dist = 0
    final_angular_dist = angle * 2 * math.pi / 360
    start_time = rospy.Time.now().to_sec()

    while current_angular_dist < final_angular_dist:
        pub.publish(vel_msg)
        time = rospy.Time.now().to_sec()
        current_angular_dist = (time - start_time) * angular_velocity
    
    # stop the robot
    vel_msg.angular.z = 0
    pub.publish(vel_msg)

def pose_changed(data:Pose):
    global initial_x, initial_y, print_pos
    if initial_x == -1:
        initial_x = data.x
        initial_y = data.y
    if print_pos:
        rospy.loginfo("The starting position is x: %f, y: %f and the ending position is x: %f, y: %f",
        initial_x,
        initial_y,
        data.x,
        data.y)
        print_pos = False

def main():
    rospy.init_node('turtle_mover', anonymous=True)
    rospy.Subscriber('turtle1/pose', Pose, pose_changed)
    rospy.sleep(3) # wait for 3 seconds

    move(1, 0.2)
    rotate(45,5)
    global print_pos
    print_pos = True
    rospy.sleep(2) 
    # sleep for 2 seconds
    # so that subscribers can listen one more time to print_pos

if __name__ == "__main__":
    main()

File: turtle_mover/src/turtle_mover.cpp
#include <ros/ros.h>
#include <geometry_msgs/Twist.h>
#include <turtlesim/Pose.h>

ros::Publisher pub;
ros::Subscriber sub;
double initial_x = -1;
double initial_y = -1;
double final_x = -1;
double final_y = -1;

const double METERS_FORWARD = 1;
bool printPos = false;

// method to move the robot by 1m
void move(double speed, double distance);

// method to rotate the robot by 45 degs
void rotate(double speed, double deg);

/**
 * @brief Subscriber callback for /turtle1/pose
 * 
 * @param pose Turtlesim pose object
 */
void callback(turtlesim::PosePtr const &pose);




int main(int argc, char **argv)
{
    ros::init(argc, argv, "turtle_mover");    
    ros::NodeHandle n;
    pub = n.advertise<geometry_msgs::Twist>("turtle1/cmd_vel", 10);
    sub = n.subscribe("/turtle1/pose",1000, callback);
    ros::Rate sleeper(0.5);
    sleeper.sleep();
    
    
    
    move(0.2,1);
    rotate(5,45);
    printPos = true;
    ros::spinOnce();
    
    return 0;
}

void move(double speed, double distance) {
    ROS_INFO("Moving robot 1m");
    geometry_msgs::Twist vel_msg;
    vel_msg.linear.x = speed;
    vel_msg.linear.y = 0;
    vel_msg.linear.z = 0;
    vel_msg.angular.x = 0;
    vel_msg.angular.y = 0;
    vel_msg.angular.z = 0;
    
    // get the current time
    double t1 = ros::Time::now().toSec();
    double current_distance = 0;

    ros::Rate loop(30);
    do {
        pub.publish(vel_msg);
        double t2 = ros::Time::now().toSec();
        current_distance = (t2-t1) * speed;
        ROS_INFO("current distance %f", current_distance);
        ros::spinOnce();
        loop.sleep();
        } while(current_distance < METERS_FORWARD);

        // stop the robot
        vel_msg.linear.x = 0;
        pub.publish(vel_msg);
}

void rotate(double speed, double angle) {
    ROS_INFO("Rotating robot 45 degs");
    double angular_speed = speed * 2 * M_PI / 360;
    geometry_msgs::Twist vel_msg;
    vel_msg.linear.x = 0;
    vel_msg.linear.y = 0;
    vel_msg.linear.z = 0;
    vel_msg.angular.x = 0;
    vel_msg.angular.y = 0;
    vel_msg.angular.z = angular_speed;

    // M_PI
    double t1 = ros::Time::now().toSec();
    double current_angle = 0;
    double final_angle = angle * 2 * M_PI / 360;
    ros::Rate loop(30);
    do {
        pub.publish(vel_msg);
        double t2 = ros::Time::now().toSec();
        current_angle = (t2-t1) * angular_speed;
        ros::spinOnce();
        loop.sleep();
    } while(current_angle < final_angle);
    // stop
    vel_msg.angular.z = 0;
    pub.publish(vel_msg);
}


void callback(turtlesim::PosePtr const &pose) {
    if (initial_x == -1) {
        initial_x = pose->x;
        initial_y = pose->y;
    }
    if (printPos) {
        ROS_INFO("The initial position is x: %f, y: %f and the final position is x: %f, y: %f",
        initial_x,
        initial_y,
        pose->x,
        pose->y);
        printPos = false;
    }
}